### PR TITLE
Replace Black with Ruff

### PR DESCRIPTION
**Pull request by tiran**
# Changes

**Which issue is resolved by this Pull Request:**
See: #739

**Description of your changes:**
Ruff is an extremely fast Python linter and code formatter, written in Rust. It is a fantastic tool that combines the features of flake8 linter, isort, Black code formatter, and (to some degree) pylint in one tool. It comes with over 800 linting rules. Ruff is not only blazing fast, it can also fix most code violation. The code fixer is a fantastic quality of life feature for developers.

Gradually introduce Ruff for linting and code formatting. Replaces Black as code formatting tool. Ruff is configured to create the same results as Black's default configuration.

More Ruff linters will be enabled in future commits.

`make verify` now performs a fast and lean check, which detects most problems except for some import errors.

```shell
$ rm -rf .tox
$ make verify
tox p -e fastlint,ruff
ruff: OK ✔ in 4.09 seconds
  fastlint: OK (13.21=setup[7.08]+cmd[6.13] seconds)
  ruff: OK (4.09=setup[3.85]+cmd[0.25] seconds)
  congratulations :) (13.49 seconds)
$ du --si -s .tox/
65M     .tox/
```

Normal `lint` downloads and creates a large virtual env, which takes several minutes:
```shell
$ tox -e lint
...
  lint: OK (178.89=setup[151.78]+cmd[27.10] seconds)
  congratulations :) (178.97 seconds)
$ du --si -s .tox/
1,6G    .tox/
```

### CODE REVIEW

**Comment by tiran** [comment]
This PR was formerly known as #740.

File: scripts/ruff.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        exitcode=$(( exitcode + $? ))
+
+        run isort --check --diff .
+        exitcode=$(( exitcode + $? ))
+        set -e
+
+        if [ $exitcode -ne 0 ]; then
+            echo "ERROR: one or more checks have failed" >&2
+            exit 3
+        fi
```
Comment (leseb) [inline scripts/ruff.sh:44]: Can we help the user a bit more? How about having an array with the failed checks and printing it with the error message? So the user knows which check failed.

Comment (tiran) [reply scripts/ruff.sh:44]: ruff and isort are already printing detailed error information to stderr. Here is an example with a missing isort header and an extra newline:

```
$ tox -e ruff -- check
ruff: install_deps> python -I -m pip install isort==5.11.5 ruff==0.3.4
ruff: commands[0]> ./scripts/ruff.sh check
RUN: 'ruff check .'
All checks passed!

RUN: 'ruff format --diff .'
--- src/instructlab/lab.py
+++ src/instructlab/lab.py
@@ -28,7 +28,6 @@
     import torch
 
 
-
 class Lab:
     """Lab object holds high-level information about ilab CLI"""
 

1 file would be reformatted, 46 files already formatted

RUN: 'isort --check --diff .'
ERROR: src/instructlab/lab.py Imports are incorrectly sorted and/or formatted.
--- src/instructlab/lab.py:before        2024-04-22 09:42:27.262042
+++ src/instructlab/lab.py:after 2024-04-22 09:42:39.052428
@@ -12,6 +12,7 @@
 import sys
 import typing
 
+# Third Party
 from click_didyoumean import DYMGroup
 from git import GitError, Repo
 from huggingface_hub import hf_hub_download
Skipped 5 files

ERROR: one or more checks have failed
ruff: exit 3 (0.31 seconds) /home/heimes/redhat/ai/instructlab/instructlab> ./scripts/ruff.sh check pid=384394
  ruff: FAIL code 3 (4.46=setup[4.16]+cmd[0.31] seconds)
  evaluation failed :( (4.98 seconds)
```

Comment (tiran) [reply scripts/ruff.sh:44]: I could add `Run 'tox -e ruff' to auto-correct all fixable errors` to the output. WDYT?

Comment (tiran) [reply scripts/ruff.sh]: The error message now points to `tox -e ruff`

**Pull request merged**